freqtrade/freqai/data_kitchen.py

```diff
@@ -1350,6 +1374,109 @@ def check_if_backtest_prediction_is_valid(
             )
             return False
 
+    def set_timerange_from_ready_models(self):
+        backtesting_timerange, \
+            assets_end_dates = (
+                self.get_timerange_and_assets_end_dates_from_ready_models(self.full_path))
+
+        self.backtest_live_models_data = {
+            "backtesting_timerange": backtesting_timerange,
+            "assets_end_dates": assets_end_dates
+            }
+        return
+
+    def get_full_models_path(self, config: Config) -> Path:
+        """
+        Returns default FreqAI model path
+        :param config: Configuration dictionary
+        """
+        freqai_config: Dict[str, Any] = config["freqai"]
+        return Path(
+            config["user_data_dir"] / "models" / str(freqai_config.get("identifier"))
+        )
+
+    def get_timerange_and_assets_end_dates_from_ready_models(
+            self, models_path: Path) -> Tuple[TimeRange, Dict[str, Any]]:
+        """
+        Returns timerange information based on a FreqAI model directory
+        :param models_path: FreqAI model path
+
+        :return: a Tuple with (Timerange calculated from directory and
+        a Dict with pair and model end training dates info)
+        """
+        all_models_end_dates = []
+        assets_end_dates: Dict[str, Any] = self.get_assets_timestamps_training_from_ready_models(
+                                           models_path)
+        for key in assets_end_dates:
+            for model_end_date in assets_end_dates[key]:
+                if model_end_date not in all_models_end_dates:
+                    all_models_end_dates.append(model_end_date)
+
+        if len(all_models_end_dates) == 0:
+            raise OperationalException(
+                'At least 1 saved model is required to '
+                'run backtest with the freqai-backtest-live-models option'
+            )
+
+        if len(all_models_end_dates) == 1:
+            logger.warning(
+                "Only 1 model was found. Backtesting will run with the "
+                "timerange from the end of the training date to the current date"
+            )
+
+        finish_timestamp = int(datetime.now(tz=timezone.utc).timestamp())
+        if len(all_models_end_dates) > 1:
+            # After last model end date, use the same period from previous model
+            # to finish the backtest
+            all_models_end_dates.sort(reverse=True)
+            finish_timestamp = all_models_end_dates[0] + \
+                (all_models_end_dates[0] - all_models_end_dates[1])
+
+        all_models_end_dates.append(finish_timestamp)
+        all_models_end_dates.sort()
+        start_date = (datetime(*datetime.fromtimestamp(min(all_models_end_dates),
+                      timezone.utc).timetuple()[:3], tzinfo=timezone.utc))
+        end_date = (datetime(*datetime.fromtimestamp(max(all_models_end_dates),
+                    timezone.utc).timetuple()[:3], tzinfo=timezone.utc))
```

Comment:
am i correct in assuming that this will trim the model start/end date to days?
so if my live model was running from 2022-10-08 23:00:00 - 2022-10-09 01:00:00 - i'd get `2022-10-08 - 2022-10-10`?

is there a particular reason for this? (backtesting can handle intra-day start times through `--timerange <epoch>-<epoch>` - though it's no longer a documented behavior...)

Reply:
currently, Freqai needs a string timerange in config (see this code: https://github.com/freqtrade/freqtrade/blob/backtest_live_models/freqtrade/freqai/data_kitchen.py#L1035). The main purpose of this code is to return this format to keeps the compatibility with existent FreqAI backtest behavior. 

But, the code will only make the predictions using the timestamp models periods. Even if the BT dataframe contains all candles, predictions and possibly trades will only happen on candles calculated based on the models timestamp. The other candles outside the period will not have the model predictions.

Reply:
@xmatthias Sorry. Forget my previous comment. I'm going to do some tests with the timerange with epochs.

Reply:
Just asking ;) for now it's good as is :+1: 

let's keep the epoch version for a later enhancement - this PR was "lying around" for long enough.